File: src/engines/addition.ts
import type { StepInput, GridCell, StepExplanation, OperationLayout } from './types';

function toDigits(n: number): number[] {
  return String(n).split('').map(Number);
}

export interface AdditionLayout {
  rows: number;
  cols: number;
  operandRows: number[];
  answerRow: number;
  carryRow: number;
  operatorCol: number;
  lineRow: number;
}

export function getAdditionLayout(operands: number[]): AdditionLayout {
  const allDigits = operands.map(n => toDigits(n));
  const maxLen = Math.max(...allDigits.map(d => d.length));
  const answerLen = maxLen + 1; // possible carry
  const cols = answerLen + 1; // +1 for operator column
  const numOperands = operands.length;
  const rows = 1 + numOperands + 1; // carry + operands + answer

  const operandRows = Array.from({ length: numOperands }, (_, i) => i + 1);

  return {
    rows,
    cols,
    carryRow: 0,
    operandRows,
    answerRow: rows - 1,
    operatorCol: 0,
    lineRow: rows - 1,
  };
}

export function generateAdditionSteps(operands: number[]): StepInput[] {
  const allDigits = operands.map(n => toDigits(n));
  const maxLen = Math.max(...allDigits.map(d => d.length));

  // Pad all to equal length
  for (const d of allDigits) {
    while (d.length < maxLen) d.unshift(0);
  }

  const layout = getAdditionLayout(operands);
  const steps: StepInput[] = [];
  let carry = 0;
  let stepIndex = 0;

  // Work right-to-left through columns
  for (let i = maxLen - 1; i >= 0; i--) {
    const columnSum = allDigits.reduce((sum, d) => sum + d[i], 0) + carry;
    const answerDigit = columnSum % 10;
    const newCarry = Math.floor(columnSum / 10);

    const col = i + 1 + (layout.cols - 1 - maxLen);

    // Step: answer digit
    const answerStepId = `step-${stepIndex++}`;
    steps.push({
      id: answerStepId,
      type: 'answer_digit',
      position: { row: layout.answerRow, col, layer: 'main' },
      correctValue: answerDigit,
      enteredValue: null,
      status: 'pending',
      compoundValue: newCarry > 0 ? columnSum : undefined,
    });

    // Step: carry digit (if needed)
    if (newCarry > 0) {
      const carryCol = col - 1;
      steps.push({
        id: `step-${stepIndex++}`,
        type: 'carry',
        position: { row: layout.carryRow, col: carryCol, layer: 'carry' },
        correctValue: newCarry,
        enteredValue: null,
        status: 'pending',
        linkedStepId: answerStepId,
      });
    }

    carry = newCarry;
  }

  // If there's a final carry that extends the answer (e.g. 999+1=1000)
  if (carry > 0) {
    steps.push({
      id: `step-${stepIndex++}`,
      type: 'answer_digit',
      position: { row: layout.answerRow, col: 1, layer: 'main' },
      correctValue: carry,
      enteredValue: null,
      status: 'pending',
    });
  }

  return steps;
}

export function getAdditionGridCells(
  operands: number[],
  steps: StepInput[],
  showHints?: boolean,
  pendingFirstDigit?: number | null
): GridCell[] {
  const allDigits = operands.map(n => toDigits(n));
  const maxLen = Math.max(...allDigits.map(d => d.length));
  const layout = getAdditionLayout(operands);

  // Pad all to equal length
  for (const d of allDigits) {
    while (d.length < maxLen) d.unshift(0);
  }

  const cells: GridCell[] = [];

  // Operator symbol on the last operand row
  const lastOperandRow = layout.operandRows[layout.operandRows.length - 1];
  cells.push({
    row: lastOperandRow,
    col: layout.operatorCol,
    layer: 'main',
    content: '+',
    editable: false,
    status: 'fixed',
  });

  // All operand digits
  for (let opIdx = 0; opIdx < operands.length; opIdx++) {
    const d = allDigits[opIdx];
    const row = layout.operandRows[opIdx];
    for (let i = 0; i < maxLen; i++) {
      const col = i + 1 + (layout.cols - 1 - maxLen);
      if (d[i] !== 0 || i > 0) {
        cells.push({
          row,
          col,
          layer: 'main',
          content: String(d[i]),
          editable: false,
          status: 'fixed',
        });
      }
    }
  }

  // Step-based cells (answer digits and carries)
  // Build a set of active answer step IDs for tentative carry lookup
  const activeAnswerStepId = steps.find(s => s.status === 'active' && s.type === 'answer_digit')?.id;

  for (const step of steps) {
    const isCompleted = step.status === 'completed';
    const isActive = step.status === 'active';

    // Check if this is a carry step linked to the currently active answer step with a pending first digit
    const isTentative = step.type === 'carry'
      && step.linkedStepId === activeAnswerStepId
      && pendingFirstDigit != null
      && step.status === 'pending';

    cells.push({
      row: step.position.row,
      col: step.position.col,
      layer: step.position.layer,
      content: isCompleted ? String(step.correctValue)
             : isTentative ? String(pendingFirstDigit)
             : (showHints && isActive) ? String(step.correctValue)
             : '',
      editable: true,
      status: isTentative ? 'tentative'
             : isActive ? 'active'
             : isCompleted ? 'completed'
             : 'pending',
      type: step.type,
      stepId: step.id,
    });
  }

  return cells;
}

export function getAdditionOperationLayout(operands: number[]): OperationLayout {
  const layout = getAdditionLayout(operands);
  const lastOperandRow = layout.operandRows[layout.operandRows.length - 1];
  return { rows: layout.rows, cols: layout.cols, lineAfterRow: lastOperandRow };
}

export function getAdditionStepExplanation(
  operands: number[],
  steps: StepInput[],
  currentStepIndex: number
): StepExplanation {
  const allDigits = operands.map(n => toDigits(n));
  const maxLen = Math.max(...allDigits.map(d => d.length));
  for (const d of allDigits) {
    while (d.length < maxLen) d.unshift(0);
  }

  if (currentStepIndex >= steps.length) {
    return { title: 'Complete!', detail: `The answer is ${operands.reduce((a, b) => a + b, 0)}.` };
  }

  const step = steps[currentStepIndex];
  const colFromRight = maxLen - (step.position.col - 1);
  const placeNames = ['ones', 'tens', 'hundreds', 'thousands', 'ten-thousands'];
  const placeName = placeNames[colFromRight - 1] || `place ${colFromRight}`;

  if (step.type === 'carry') {
    const linkedStep = steps.find(s => s.id === step.linkedStepId);
    if (linkedStep) {
      const prevCol = maxLen - (linkedStep.position.col - 1);
      const prevPlace = placeNames[prevCol - 1] || `place ${prevCol}`;
      return {
        title: `Write the carry`,
        detail: `The ${prevPlace} column added up to 10 or more, so carry ${step.correctValue} to the ${placeName} column.`,
      };
    }
    return { title: 'Write the carry', detail: `Carry ${step.correctValue} to the next column.` };
  }

  // answer_digit
  const columnDigits = allDigits.map(d => d[step.position.col - 1]);
  // Check if there was a carry into this column
  const prevCarryStep = steps.find(
    s => s.type === 'carry' && s.status === 'completed' && s.position.col === step.position.col
  );
  const carryIn = prevCarryStep ? prevCarryStep.correctValue : 0;
  const parts = columnDigits.filter(d => d !== undefined);
  const sumStr = [...parts, ...(carryIn > 0 ? [carryIn] : [])].join(' + ');

  return {
    title: `Add the ${placeName} column`,
    detail: `${sumStr} = ${step.correctValue}${step.compoundValue !== undefined && step.compoundValue >= 10 ? ` (write ${step.correctValue}, carry ${Math.floor(step.compoundValue / 10)})` : ''}`,
  };
}

File: go-engines/addition.go
package engines

import "fmt"

// AdditionLayout describes the grid layout for an addition problem.
type AdditionLayout struct {
	Rows        int
	Cols        int
	OperandRows []int
	AnswerRow   int
	CarryRow    int
	OperatorCol int
	LineRow     int
}

// GetAdditionLayout computes the grid layout for the given operands.
func GetAdditionLayout(operands []int) AdditionLayout {
	allDigits := make([][]int, len(operands))
	maxLen := 0
	for i, n := range operands {
		allDigits[i] = toDigits(n)
		if len(allDigits[i]) > maxLen {
			maxLen = len(allDigits[i])
		}
	}
	answerLen := maxLen + 1 // possible carry
	cols := answerLen + 1   // +1 for operator column
	numOperands := len(operands)
	rows := 1 + numOperands + 1 // carry + operands + answer

	operandRows := make([]int, numOperands)
	for i := range operandRows {
		operandRows[i] = i + 1
	}

	return AdditionLayout{
		Rows:        rows,
		Cols:        cols,
		CarryRow:    0,
		OperandRows: operandRows,
		AnswerRow:   rows - 1,
		OperatorCol: 0,
		LineRow:     rows - 1,
	}
}

// GenerateAdditionSteps generates the step-by-step solution for an addition problem.
func GenerateAdditionSteps(operands []int) []StepInput {
	allDigits := make([][]int, len(operands))
	maxLen := 0
	for i, n := range operands {
		allDigits[i] = toDigits(n)
		if len(allDigits[i]) > maxLen {
			maxLen = len(allDigits[i])
		}
	}

	// Pad all to equal length
	for i := range allDigits {
		for len(allDigits[i]) < maxLen {
			allDigits[i] = append([]int{0}, allDigits[i]...)
		}
	}

	layout := GetAdditionLayout(operands)
	var steps []StepInput
	stepIndex := 0
	carry := 0

	// Work right-to-left through columns
	for i := maxLen - 1; i >= 0; i-- {
		columnSum := carry
		for _, d := range allDigits {
			columnSum += d[i]
		}
		answerDigit := columnSum % 10
		newCarry := columnSum / 10

		col := i + 1 + (layout.Cols - 1 - maxLen)

		answerStepID := fmt.Sprintf("step-%d", stepIndex)
		stepIndex++

		step := StepInput{
			ID:           answerStepID,
			Type:         AnswerDigit,
			Position:     GridPosition{Row: layout.AnswerRow, Col: col, Layer: LayerMain},
			CorrectValue: answerDigit,
			Status:       Pending,
		}
		if newCarry > 0 {
			step.CompoundValue = intPtr(columnSum)
		}
		steps = append(steps, step)

		// Carry digit if needed
		if newCarry > 0 {
			carryCol := col - 1
			steps = append(steps, StepInput{
				ID:           fmt.Sprintf("step-%d", stepIndex),
				Type:         Carry,
				Position:     GridPosition{Row: layout.CarryRow, Col: carryCol, Layer: LayerCarry},
				CorrectValue: newCarry,
				Status:       Pending,
				LinkedStepID: answerStepID,
			})
			stepIndex++
		}

		carry = newCarry
	}

	// Final carry that extends the answer
	if carry > 0 {
		steps = append(steps, StepInput{
			ID:           fmt.Sprintf("step-%d", stepIndex),
			Type:         AnswerDigit,
			Position:     GridPosition{Row: layout.AnswerRow, Col: 1, Layer: LayerMain},
			CorrectValue: carry,
			Status:       Pending,
		})
	}

	return steps
}

// GetAdditionGridCells returns the grid cells for rendering an addition problem.
func GetAdditionGridCells(operands []int, steps []StepInput, showHints bool, pendingFirstDigit *int) []GridCell {
	allDigits := make([][]int, len(operands))
	maxLen := 0
	for i, n := range operands {
		allDigits[i] = toDigits(n)
		if len(allDigits[i]) > maxLen {
			maxLen = len(allDigits[i])
		}
	}

	layout := GetAdditionLayout(operands)

	// Pad all to equal length
	for i := range allDigits {
		for len(allDigits[i]) < maxLen {
			allDigits[i] = append([]int{0}, allDigits[i]...)
		}
	}

	var cells []GridCell

	// Operator symbol on last operand row
	lastOperandRow := layout.OperandRows[len(layout.OperandRows)-1]
	cells = append(cells, GridCell{
		Row:     lastOperandRow,
		Col:     layout.OperatorCol,
		Layer:   LayerMain,
		Content: "+",
		Status:  Fixed,
	})

	// All operand digits
	for opIdx := 0; opIdx < len(operands); opIdx++ {
		d := allDigits[opIdx]
		row := layout.OperandRows[opIdx]
		for i := 0; i < maxLen; i++ {
			col := i + 1 + (layout.Cols - 1 - maxLen)
			if d[i] != 0 || i > 0 {
				cells = append(cells, GridCell{
					Row:     row,
					Col:     col,
					Layer:   LayerMain,
					Content: fmt.Sprintf("%d", d[i]),
					Status:  Fixed,
				})
			}
		}
	}

	// Find active answer step ID
	activeAnswerStepID := ""
	for _, s := range steps {
		if s.Status == Active && s.Type == AnswerDigit {
			activeAnswerStepID = s.ID
			break
		}
	}

	// Step-based cells
	for _, step := range steps {
		isCompleted := step.Status == Completed
		isActive := step.Status == Active

		isTentative := step.Type == Carry &&
			step.LinkedStepID == activeAnswerStepID &&
			pendingFirstDigit != nil &&
			step.Status == Pending

		content := ""
		var status CellStatus
		if isCompleted {
			content = fmt.Sprintf("%d", step.CorrectValue)
			status = CellCompleted
		} else if isTentative {
			content = fmt.Sprintf("%d", *pendingFirstDigit)
			status = Tentative
		} else if showHints && isActive {
			content = fmt.Sprintf("%d", step.CorrectValue)
			status = CellActive
		} else if isActive {
			status = CellActive
		} else {
			status = CellPending
		}

		cells = append(cells, GridCell{
			Row:      step.Position.Row,
			Col:      step.Position.Col,
			Layer:    step.Position.Layer,
			Content:  content,
			Editable: true,
			Status:   status,
			Type:     step.Type,
			StepID:   step.ID,
		})
	}

	return cells
}

// GetAdditionOperationLayout returns the OperationLayout for an addition problem.
func GetAdditionOperationLayout(operands []int) OperationLayout {
	layout := GetAdditionLayout(operands)
	lastOperandRow := layout.OperandRows[len(layout.OperandRows)-1]
	return OperationLayout{Rows: layout.Rows, Cols: layout.Cols, LineAfterRow: &lastOperandRow}
}

// GetAdditionStepExplanation returns a human-readable explanation for the current step.
func GetAdditionStepExplanation(operands []int, steps []StepInput, currentStepIndex int) StepExplanation {
	allDigits := make([][]int, len(operands))
	maxLen := 0
	for i, n := range operands {
		allDigits[i] = toDigits(n)
		if len(allDigits[i]) > maxLen {
			maxLen = len(allDigits[i])
		}
	}
	for i := range allDigits {
		for len(allDigits[i]) < maxLen {
			allDigits[i] = append([]int{0}, allDigits[i]...)
		}
	}

	if currentStepIndex >= len(steps) {
		sum := 0
		for _, o := range operands {
			sum += o
		}
		return StepExplanation{
			Title:  "Complete!",
			Detail: fmt.Sprintf("The answer is %d.", sum),
		}
	}

	step := steps[currentStepIndex]
	colFromRight := maxLen - (step.Position.Col - 1)
	placeNames := []string{"ones", "tens", "hundreds", "thousands", "ten-thousands"}
	placeName := fmt.Sprintf("place %d", colFromRight)
	if colFromRight-1 >= 0 && colFromRight-1 < len(placeNames) {
		placeName = placeNames[colFromRight-1]
	}

	if step.Type == Carry {
		for _, s := range steps {
			if s.ID == step.LinkedStepID {
				prevCol := maxLen - (s.Position.Col - 1)
				prevPlace := fmt.Sprintf("place %d", prevCol)
				if prevCol-1 >= 0 && prevCol-1 < len(placeNames) {
					prevPlace = placeNames[prevCol-1]
				}
				return StepExplanation{
					Title:  "Write the carry",
					Detail: fmt.Sprintf("The %s column added up to 10 or more, so carry %d to the %s column.", prevPlace, step.CorrectValue, placeName),
				}
			}
		}
		return StepExplanation{
			Title:  "Write the carry",
			Detail: fmt.Sprintf("Carry %d to the next column.", step.CorrectValue),
		}
	}

	// answer_digit
	columnDigits := make([]int, 0, len(allDigits))
	for _, d := range allDigits {
		columnDigits = append(columnDigits, d[step.Position.Col-1])
	}

	carryIn := 0
	for _, s := range steps {
		if s.Type == Carry && s.Status == Completed && s.Position.Col == step.Position.Col {
			carryIn = s.CorrectValue
			break
		}
	}

	sumStr := ""
	for i, d := range columnDigits {
		if i > 0 {
			sumStr += " + "
		}
		sumStr += fmt.Sprintf("%d", d)
	}
	if carryIn > 0 {
		sumStr += fmt.Sprintf(" + %d", carryIn)
	}

	detail := fmt.Sprintf("%s = %d", sumStr, step.CorrectValue)
	if step.CompoundValue != nil && *step.CompoundValue >= 10 {
		detail += fmt.Sprintf(" (write %d, carry %d)", step.CorrectValue, *step.CompoundValue/10)
	}

	return StepExplanation{
		Title:  fmt.Sprintf("Add the %s column", placeName),
		Detail: detail,
	}
}
